pandas/types/cast.py

```diff
@@ -322,8 +325,52 @@ def infer_dtype_from_scalar(val, pandas_dtype=False):
         whether to infer dtype including pandas extension types.
         If False, scalar belongs to pandas extension types is inferred as
         object
+    downcast : bool, default False
+        If True, downcast float and integer types to the smallest width
+        type that can hold `val`.
+
+    .. versionadded:: 0.20.0
+    allow_uint : bool, default False
+        If True and `downcast` is True, non-negative integers will be
+        downcast to smallest width unsigned integer type that can hold
+        them. Otherwise, signed types are always downcast to signed types
+        and the same for unsigned types.
+
+    .. versionadded:: 0.20.0
     """
 
+    def _downcast_dtype(dtype, val):
+        if 'float' in str(dtype):
+            if ((val > np.finfo(np.float16).min and
+                    val < np.finfo(np.float16).max) or val is np.nan):
+                return np.float16
+            elif (val > np.finfo(np.float32).min and
+                    val < np.finfo(np.float32).max):
```

Comment:
actually you could prob take this *exact* code from ``to_numeric`` and put it in a function and call it here (wrap the scalar), and replace in ``to_numeric``.
```
    if downcast is not None and is_numeric_dtype(values):
        typecodes = None

        if downcast in ('integer', 'signed'):
            typecodes = np.typecodes['Integer']
        elif downcast == 'unsigned' and np.min(values) >= 0:
            typecodes = np.typecodes['UnsignedInteger']
        elif downcast == 'float':
            typecodes = np.typecodes['Float']

            # pandas support goes only to np.float32,
            # as float dtypes smaller than that are
            # extremely rare and not well supported
            float_32_char = np.dtype(np.float32).char
            float_32_ind = typecodes.index(float_32_char)
            typecodes = typecodes[float_32_ind:]

        if typecodes is not None:
            # from smallest to largest
            for dtype in typecodes:
                if np.dtype(dtype).itemsize <= values.dtype.itemsize:
                    values = maybe_downcast_to_dtype(values, dtype)

                    # successful conversion
                    if values.dtype == dtype:
                        break
```